samples/oauth_rails_vethealth/app/controllers/session_controller.rb

```diff
@@ -0,0 +1,125 @@
+    if params[:code].nil?
+      if params[:error].present?
+        flash.alert = "Callback received \"#{params['error']}\" error: #{params['error_description']}"
+      else
+        flash.notice = "The callback page is for oauth responses, please login first."
+      end
+      redirect_to(login_path) and return
+    end
+    @verified_state = params[:state].to_i == session[:login_time]
+    @body = {
+      code: params[:code],
+      state: params[:state],
+      grant_type: 'authorization_code',
+      redirect_uri: 'http://localhost:3000/callback'
+    }
+    @post_url = 'https://dev-api.va.gov/oauth2/token'
+    auth = { username: ENV['va_developer_client_id'], password: ENV['va_developer_client_secret'] }
+    @auth = auth.dup
+    @auth[:password] = "<Client Secret found in application.yml>"
+    @session = nil
+
+    # set oauth_response and oauth_response_code
+    if session[:oauth_code]
+      # we are reloading the page so we don't want to make a stale request
+      @oauth_response_code = session[:oauth_code]
+      @oauth_response = 
+        if session[:oauth_response].is_a? Hash
+          # response was not saved as a Session model
+          if @oauth_response_code == 200
+            # 200 OK means it was a successful request but session was found invalid
+            attributes = Session.attributes_from_oauth(oauth_post)
+            @session = Session.new(attributes)
+          end
+          session[:oauth_response]
+        else
+          # session[:oauth_response] is an array of keys, so get data from saved session
+          @session = Session.find(session[:id])
+          hide_sensitive_data(@session.attributes.reject { |k,v| %w(id created_at updated_at).include?(k) })
+        end
+      @session.validate_session(session) if @session
+    else
+      oauth_post = HTTParty.post(@post_url, { basic_auth: auth, body: @body })
+      @oauth_response = hide_sensitive_data(oauth_post.to_h)
+      @oauth_response_code = oauth_post.code
+      # save oauth_post to support refreshing this page
+      session[:oauth_code] = @oauth_response_code
+      session[:oauth_response] = @oauth_response
+
+      if oauth_post.ok?
+        attributes = Session.attributes_from_oauth(oauth_post)
+        @session = Session.new(attributes)
+        if @session.validate_session(session)
+          @session.save!
+          session[:id] = @session.id
+          session[:oauth_response] = attributes.keys # don't store the sensitive data in the session
+        end
+      end
+    end
+  end
+
+  def logout
+    if session[:id]
+      sesh = Session.find(session[:id])
+      if Time.zone.now < sesh.expires_at
+        sesh.expires_at = Time.zone.now
+        sesh.save!
+      end
+    end
+    redirect_to login_path
+  end
+
+  def show
+    @session =
+      if params[:id].blank?
+        Session.find(session[:id])
+      else
+        Session.find(params[:id])
+      end
+  end
+
+private
+  def hide_sensitive_data(response_hash)
```

Comment:
Could you explain why we are hiding this data in the response?

Reply:
The `access_token` because if this was captured someone could impersonate requests to the APIs, so I decided to hide it as a best practice sort of thing.  I went back and forth on wether or not to show the JWT.  It's got the name in there so it would be PII (if it wasn't a test user), but I think I ended up showing it on the `session#show` page.